src/txrequest.cpp

```diff
@@ -0,0 +1,724 @@
+        return ByTxHashView{ann.m_txhash, ann.m_state, prio};
+    }
+};
+
+enum class WaitState {
+    //! Used for announcements that need efficient testing of "is their timestamp in the future?".
+    FUTURE_EVENT,
+    //! Used for announcements whose timestamp is not relevant.
+    NO_EVENT,
+    //! Used for announcements that need efficient testing of "is their timestamp in the past?".
+    PAST_EVENT,
+};
+
+WaitState GetWaitState(const Announcement& ann)
+{
+    if (ann.IsWaiting()) return WaitState::FUTURE_EVENT;
+    if (ann.IsSelectable()) return WaitState::PAST_EVENT;
+    return WaitState::NO_EVENT;
+}
+
+// The ByTime index is sorted by (wait_state, time).
+//
+// All announcements with a timestamp in the future can be found by iterating the index forward from the beginning.
+// All announcements with a timestamp in the past can be found by iterating the index backwards from the end.
+//
+// Uses:
+// * Finding CANDIDATE_DELAYED announcements whose reqtime has passed, and REQUESTED announcements whose expiry has
+//   passed.
+// * Finding CANDIDATE_READY/BEST announcements whose reqtime is in the future (when the clock time went backwards).
+struct ByTime {};
+using ByTimeView = std::pair<WaitState, std::chrono::microseconds>;
+struct ByTimeViewExtractor
+{
+    using result_type = ByTimeView;
+    result_type operator()(const Announcement& ann) const
+    {
+        return ByTimeView{GetWaitState(ann), ann.m_time};
+    }
+};
+
+/** Data type for the main data structure (Announcement objects with ByPeer/ByTxHash/ByTime indexes). */
+using Index = boost::multi_index_container<
+    Announcement,
+    boost::multi_index::indexed_by<
+        boost::multi_index::ordered_unique<boost::multi_index::tag<ByPeer>, ByPeerViewExtractor>,
+        boost::multi_index::ordered_non_unique<boost::multi_index::tag<ByTxHash>, ByTxHashViewExtractor>,
+        boost::multi_index::ordered_non_unique<boost::multi_index::tag<ByTime>, ByTimeViewExtractor>
+    >
+>;
+
+/** Helper type to simplify syntax of iterator types. */
+template<typename Tag>
+using Iter = typename Index::index<Tag>::type::iterator;
+
+/** Per-peer statistics object. */
+struct PeerInfo {
+    size_t m_total = 0; //!< Total number of announcements for this peer.
+    size_t m_completed = 0; //!< Number of COMPLETED announcements for this peer.
+    size_t m_requested = 0; //!< Number of REQUESTED announcements for this peer.
+};
+
+/** Per-txhash statistics object. Only used for sanity checking. */
+struct TxHashInfo
+{
+    //! Number of CANDIDATE_DELAYED announcements for this txhash.
+    size_t m_candidate_delayed = 0;
+    //! Number of CANDIDATE_READY announcements for this txhash.
+    size_t m_candidate_ready = 0;
+    //! Number of CANDIDATE_BEST announcements for this txhash (at most one).
+    size_t m_candidate_best = 0;
+    //! Number of REQUESTED announcements for this txhash.
+    size_t m_requested = 0;
```

Comment:
```suggestion
    //! Number of CANDIDATE_BEST announcements for this txhash (at most one).
    size_t m_candidate_best = 0;
    //! Number of REQUESTED announcements for this txhash (at most one).
    size_t m_requested = 0;
```

This is checked in https://github.com/bitcoin/bitcoin/blob/bf3f99291b8f207a809d6fd3309ef99eb3711388/src/txrequest.cpp#L324-L325

Reply:
Done.